fix: skip `proc.cwd()` when `cwd` is given

scarf005 opened the pull request:
<!--
  Please check the needed checkboxes ([ ] -> [x]). Leave the
  comments as they are, they won’t show on GitHub.
  We are excited about pull requests, but please try to limit the scope, provide
  a general description of the changes, and remember, it’s up to you to convince
  us to land it.
-->

### Initial checklist

*   [x] I read the support docs <!-- https://github.com/vfile/.github/blob/main/support.md -->
*   [x] I read the contributing guide <!-- https://github.com/vfile/.github/blob/main/contributing.md -->
*   [x] I agree to follow the code of conduct <!-- https://github.com/vfile/.github/blob/main/code-of-conduct.md -->
*   [x] I searched issues and couldn’t find anything (or linked relevant results below) <!-- https://github.com/search?q=user%3Avfile&type=Issues -->
*   [x] If applicable, I’ve added docs and tests

### Description of changes

make `VFile` skip unnecessary filesystem access for `process.cwd()` if `cwd` is already given.

due to how `VFile.cwd` works:

1. set to `process.cwd()` (`/` if browser)
2. override to `options.cwd` if exists as stated in https://github.com/vfile/vfile/pull/15#issuecomment-272976878

VFile will always request filesystem access even if it's not needed.

<!--do not edit: pr-->

wooorm commented:
Heya!

Let me try to fix the build issue.

> VFile will always request filesystem access even if it's unnecessary.

This project should be usable. `proc` redirects to a file that depends on conditions. Only explicitly in Node does one get `process`.

Why are you/Deno specifically using a `node` condition?

> import_map.json

Right, you specifically choose to use Node APIs. Why?

scarf005 commented:
> Only explicitly in Node does one get process.

I was using `import { VFile } from "npm:vfile@6.0.1"` out of habit, and the [`npm:`](https://docs.deno.com/runtime/manual/node/) specifier polyfilled node `process.cwd()` (`Deno.cwd()`).

changing import to `import { VFile } from "https://esm.sh/vfile@6.0.1"`, does resolve to browser-based module.

> Why are you/Deno specifically using a node condition?

to recap:

- `npm:vfile`: node polyfill, for server
- `https://esm.sh/vfile`: for browser

I think it makes sense to choose node condition depending on how the library is used. The concern I have right now would be

> Right, you specifically choose to use Node APIs. Why?

> Also, does Deno expect users to map every imported thing to a file on the file system? These internals specifically contain do-not-use. I was not aware of that. That does not sound good.

Oh, most wouldn't, please ignore this one, I experimented running repo code as-is while reproducing the issue and had to do manual mapping myself.

> > VFile will always request filesystem access even if it's unnecessary.
>
> This project should be usable. 

I'm sorry, could you elaborate further? Would it mean vfile should always request cwd even if it's explicitly given in constructor?

wooorm commented:
Ah okay, so the import map has to do with trying to make a small repro, good!

> I'm sorry, could you elaborate further? Would it mean vfile should always request cwd even if it's explicitly given in constructor?

Nope!

First: a `cwd` (and `path`) is really useful. There are many plugins and other things that will look for `package.json`s or `.git` or whatnot relating to the file. `path` can be relative (in which case it relates to `cwd`). And `cwd` (where a programmer runs code from) is also very useful.

But `process.cwd` is a Node thing. So only if Node is explicitly specified, will `process.cwd` be used:

https://github.com/vfile/vfile/blob/de87feae24b65cd21975022636b70b0ad7331342/package.json#L36-L50

*Everything else* gets the `default` condition. And so, gets:

https://github.com/vfile/vfile/blob/de87feae24b65cd21975022636b70b0ad7331342/lib/minproc.browser.js#L7

scarf005 commented:
sorry, I'm getting even more confused. My point was: 

> VFile constructor shouldn't request `cwd` via `process.cwd()` if `cwd` was explicitly given from construtor (opt.cwd)

I think this applies to not only deno but in node.

wooorm commented:
Maybe. But:
- this shouldn’t be something you notice, you are not using Node
- which cwd value are you setting? `cwd`s are quite useful, perhaps we can improve `cwd`s on deno
- how are you getting a relevant `cwd`, perhaps on deno we can use those cwd values you somehow find

scarf005 commented:
> make VFile skip unnecessary filesystem access for process.cwd() if cwd is already given.

I updated the PR description to remove relations with deno, as i changed my position to optimization (removing unneeded system call) they were less relevant.

wooorm commented:
Alright, done, slightly different mechanism but mainly wanted to add a comment!
